pkg/target/manager.go

```diff
@@ -280,79 +234,186 @@ func (m *managerImpl) validateSeed(ctx context.Context, seed *gardencorev1beta1.
 }
 
 func (m *managerImpl) TargetShoot(ctx context.Context, shootName string) error {
+	return m.patchTargetWithValues(ctx, "", "", "", "", shootName)
+}
+
+func (m *managerImpl) UnsetTargetShoot() (string, error) {
+	currentTarget, err := m.CurrentTarget()
+	if err != nil {
+		return "", fmt.Errorf("failed to get current target: %v", err)
+	}
+
+	targetedName := currentTarget.ShootName()
+	if targetedName != "" {
+		return targetedName, m.patchTarget(func(t *targetImpl) error {
+			t.Shoot = ""
+
+			return nil
+		})
+	}
+
+	return "", errors.New("no shoot targeted")
+}
+
+func (m *managerImpl) TargetMatchPattern(ctx context.Context, value string) error {
+	tm, err := m.config.MatchPattern(value)
+	if err != nil {
+		return fmt.Errorf("error occurred while trying to match value: %v", err)
+	}
+
+	if tm == nil {
+		return errors.New("the provided value does not match any pattern")
+	}
+
+	return m.patchTargetWithValues(ctx, tm.Garden, tm.Project, tm.Namespace, "", tm.Shoot)
+}
+
+func (m *managerImpl) patchTargetWithValues(ctx context.Context, gardenName string, projectName string, namespace string, seedName string, shootName string) error {
+	target, err := m.createValidTarget(ctx, gardenName, projectName, namespace, seedName, shootName)
+	if err != nil {
+		return err
+	}
+
+	if target == nil {
+		return errors.New("could not create valid target")
+	}
+
 	return m.patchTarget(func(t *targetImpl) error {
-		if t.Garden == "" {
-			return ErrNoGardenTargeted
+		t.Garden = target.GardenName()
+		t.Project = target.ProjectName()
+		t.Seed = target.SeedName()
+		t.Shoot = target.ShootName()
+
+		return nil
+	})
+}
+
+func (m *managerImpl) createValidTarget(ctx context.Context, gardenName string, projectName string, namespace string, seedName string, shootName string) (Target, error) {
+	currentTarget, err := m.CurrentTarget()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current target: %v", err)
+	}
+
+	if gardenName != "" {
+		gardenName, err = m.config.FindGarden(gardenName)
+		if err != nil {
+			return nil, err
 		}
+	} else {
+		// if newTarget does not have a Garden name set, we need to get it from the current
+		// target to be able to create a client
+		gardenName = currentTarget.GardenName()
+	}
 
-		gardenClient, err := m.clientForGarden(t.Garden)
+	if gardenName == "" {
+		return nil, ErrNoGardenTargeted
+	}
+
+	if namespace == "" && projectName == "" && seedName == "" && shootName == "" {
+		// return immediately if only garden is targeted, no need to create a client in this case
+		newTarget := NewTarget(gardenName, projectName, seedName, shootName)
+		return newTarget, nil
+	}
```

Comment:
do we really need this optimisation? I think this code is error prone in case the targeting is extended in the future

Reply:
It is not required. I implemented it for to reasons: First is that it optimizes the code as the `gardenClient` is not (unnecessarily) created. The second reason is that we would also need to adapt the test as it (currently) does not expect a client to be created and throws an error. We can change it though